core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -40,38 +41,70 @@ internal actual fun <T> createParametrizedCache(factory: (KClass<Any>, List<KTyp
+    override fun get(key: KClass<Any>): KSerializer<T>? {
+        return classValue
+            .getOrSet(key.java) { CacheEntry(compute(key)) }
+            .serializer
+    }
 }
 
+/**
+ * A class that combines the capabilities of ClassValue and SoftReference.
+ * Softly binds the calculated value to the specified class.
+ *
+ * [SoftReference] used to prevent class loaders from leaking,
+ * since the value can transitively refer to an instance of type [Class], this may prevent the loader from
+ * being collected during garbage collection.
+ *
+ * In the first calculation the value is cached, every time [getOrSet] is called, a pre-calculated value is returned.
+ *
+ * However, the value can be collected during garbage collection (thanks to [SoftReference])
+ * - in this case, when trying to call the [getOrSet] function, the value will be calculated again and placed in the cache.
+ *
+ * An important requirement for a function generating a value is that it must be stable, so that each time it is called for the same class, the function returns similar values.
+ * In the case of serializers, these should be instances of the same class filled with equivalent values.
+ */
 @SuppressAnimalSniffer
-private class ClassValueWrapper<T>(private val compute: (KClass<*>) -> KSerializer<T>?): ClassValue<CacheEntry<T>>() {
-    /*
-     * Since during the computing of the value for the `ClassValue` entry, we do not know whether a nullable
-     *  serializer is needed, so we may need to differentiate nullable/non-null  caches by a level higher
-     */
-    override fun computeValue(type: Class<*>): CacheEntry<T> {
-        return CacheEntry(compute(type.kotlin))
+private class ClassValueReferences<T> : ClassValue<MutableSoftReference<T>>() {
+    override fun computeValue(type: Class<*>): MutableSoftReference<T> {
+        return MutableSoftReference()
+    }
+
+    inline fun getOrSet(key: Class<*>, crossinline factory: () -> T): T {
+        val ref: MutableSoftReference<T> = get(key)
+
+        ref.reference.get()?.let { return it }
+        // go to the slow path and create serializer with blocking, also wrap factory block
+        return createAndSet(ref) { factory() }
     }
-}
 
-private class ClassValueParametrizedCache<T>(private val compute: (KClass<Any>, List<KType>) -> KSerializer<T>?) : ParametrizedSerializerCache<T> {
-    private val classValue = ParametrizedClassValueWrapper<T>()
+    @Synchronized
+    private fun createAndSet(ref: MutableSoftReference<T>, factory: () -> T): T {
+        val value = factory()
+        ref.reference = SoftReference(value)
+        return value
+    }
+}
 
-    override fun get(key: KClass<Any>, types: List<KType>): Result<KSerializer<T>?> =
-        classValue[key.java].computeIfAbsent(types) { compute(key, types) }
+/**
+ * Wrapper over `SoftReference`, used  to store a mutable value.
+ */
+private class MutableSoftReference<T> {
+    @Volatile
```

Comment:
I'm not sure if this is a good idea, because it's impossible to call `remove` with a condition and outside of critical section.

For example, if several threads simultaneously called `get` and received the same reference to `SoftReference`:
- first thread saw that the reference in the `SoftReference` is null, called `remove`, then calls `get` and calculate the new value of `SoftReference`.
- at this moment, the second thread also checked `SoftReference`, because it refers to the old version - saw null there and also called `remove`, re-calls `get` (to calculate the new value)

With great competition, this can actively happen and besides, two same serializers can be created, which can also be cached (for example, inside parameterized serializers).

Reply:
However, I agree, in this place `@Volatile` is useless and more confusing, because this property should be used only in the synchronized block.

Reply:
But I can call `remove` inside my synchronized block after an additional check for null. WDYT?

Reply:
Yeah, you are right, we can accidentally remove fresh reference, so it doesn't fit well